### PR TITLE
feat: bounds for ser/de derive and schema_params for schema derive attributes

### PR DESCRIPTION
**Pull request by dj8yfo**
This is logical continuation/conclusion of #178 . 

Pr adds possibility to override bounds for BorshSerialize and BorshDeserialize in order to allow not to keep
bounds on type parameters in struct definition itself and fix complex cases, when derive hasn't figured 
the right bounds on type parameters automatically, while still not having to implement the traits manually.

```rust
/// `T: PartialOrd` is injected into derived trait implementation via field bound to avoid having this restriction on
/// the struct itself
#[cfg(hash_collections)]
#[derive(BorshSerialize)]
struct C1<T, U> {
    a: String,
    #[borsh(bound(serialize = 
        "T: borsh::ser::BorshSerialize + PartialOrd,
         U: borsh::ser::BorshSerialize"))]
    b: HashMap<T, U>,
}

/// `T: PartialOrd + Hash + Eq` is injected into derived trait implementation via field bound to avoid having this restriction on
/// the struct itself
#[allow(unused)]
#[cfg(hash_collections)]
#[derive(BorshDeserialize)]
struct C2<T, U> {
    a: String,
    #[borsh(bound(
        deserialize = 
        "T: PartialOrd + Hash + Eq + borsh::de::BorshDeserialize,
         U: borsh::de::BorshDeserialize"
    ))]
    b: HashMap<T, U>,
}

/// implicit derived `core::default::Default` bounds on `K` and `V` are dropped by empty bound
/// specified, as `HashMap` has its own `Default` implementation
#[cfg(hash_collections)]
#[derive(BorshDeserialize)]
struct G3<K, V, U>(
    #[borsh_skip]
    #[borsh(bound(deserialize = ""))]
    HashMap<K, V>,
    U,
);
// derive here figures the bound erroneously as `T: borsh::ser::BorshSerialize`
#[derive(BorshSerialize)]
struct ParametrizedWronDerive<T, V>
where
    T: TraitName,
{
    #[borsh(bound(serialize = "<T as TraitName>::Associated: borsh::ser::BorshSerialize"))]
    field: <T as TraitName>::Associated,
    another: V,
}

```
Pr also adds ` #[borsh(schema(params = ...))]` attribute, which may be populated by a list of entries. `order_param => override_type`. 
Such an entry instructs `BorshSchema` derive to:

1. add `override_type` to types, bounded by `borsh::BorshSchema` in implementation block.
2. add `<override_type>::declaration()` to parameters vector in `fn declaration()` method of `BorshSchema` trait that is being derived.
3. the `order_param` is required to establish the same order in parameters vector (2.) as that of type parameters in generics of type, that `BorshSchema` is derived for.
4. entries, specified for a field, together replace whatever would've been derived automatically for 1. and 2. .

```rust
    // derive here figures the bound erroneously as `T: borsh::BorshSchema`
    // `#[borsh(schema(params = "..."))]` attribute replaces it with <T as TraitName>::Associated: borsh::BorshSchema`
    #[derive(borsh::BorshSchema)]
    struct Parametrized<V, T>
    where
        T: TraitName,
    {
        #[borsh(schema(params = "T => <T as TraitName>::Associated"))]
        field: <T as TraitName>::Associated,
        another: V,
    }
```

### CODE REVIEW

File: borsh-schema-derive-internal/src/snapshots/borsh_schema_derive_internal__enum_schema__tests__complex_enum_generics_borsh_skip_named_field.snap [hunk 1]
```diff
@@ -2,12 +2,15 @@
...
-impl<C, W> borsh::BorshSchema for A<C, W>
+impl<W, U, C> borsh::BorshSchema for A<W, U, C>
 where
+    U: borsh::BorshSchema,
     C: borsh::BorshSchema,
 {
     fn declaration() -> borsh::schema::Declaration {
-        let params = borsh::__private::maybestd::vec![< C > ::declaration()];
+        let params = borsh::__private::maybestd::vec![
+            < U > ::declaration(), < C > ::declaration()
```
Comment (dj8yfo) [inline borsh-schema-derive-internal/src/snapshots/borsh_schema_derive_internal__enum_schema__tests__complex_enum_generics_borsh_skip_named_field.snap:12]: the order has been changed to follow not alphabetic order of types transcription, but 
the original order of generics declaration in this [commit](https://github.com/near/borsh-rs/pull/180/commits/1f1a8f4be70fa7bc8bc02748be32b331c1eb3903)

**Comment by dj8yfo** [comment]
Despite work in this [commit](https://github.com/near/borsh-rs/pull/180/commits/4c0d41f3be6b622997eb338fc8ec5762b0007ea2) on filtering relevant `WherePredicate`-s, 
lifetimes in generics aren't filtered to according to their actual usage in enum variants, so following example doesn't work:
```rust
fn test_enum_with_lifetimes() {
    
    #[allow(unused)]
    #[derive(borsh::BorshSchema)]
    enum EnumParametrized<'a, 'b, T, V>
    {
        A {
            x: BTreeMap<u16, &'a T>,
        },
        B(&'b V, u16),
    }
}
```
```bash
error[E0392]: parameter `'b` is never used
   --> borsh/tests/test_schema_enums.rs:359:31
    |
359 |     enum EnumParametrized<'a, 'b, T, V>
    |                               ^^ unused parameter
    |
    = help: consider removing `'b`, referring to it in a field, or using a marker such as `PhantomData`
                                                                                                                                                                                                                      
error[E0392]: parameter `'a` is never used
   --> borsh/tests/test_schema_enums.rs:359:27
    |
359 |     enum EnumParametrized<'a, 'b, T, V>
    |                           ^^ unused parameter
    |
    = help: consider removing `'a`, referring to it in a field, or using a marker such as `PhantomData`                                                                                                                                                                                                                      
```

One way to amend this might be to add yet one more attribute to hint relevant generics to derive:
```rust
#[test]
fn test_enum_with_lifetimes() {
    
    #[allow(unused)]
    #[derive(borsh::BorshSchema)]
    enum EnumParametrized<'a, 'b, T, V>
    {
        #[borsh(schema(variant_generics = "'a, T"))]
        A {
            x: BTreeMap<u16, &'a T>,
        },
        #[borsh(schema(variant_generics = "'b, V"))]
        B(&'b V, u16),
    }
}
```
Leaving it as is as of now (not adding `borsh(schema(variant_generics = "..."))`) as `borsh-v0.11.0` isn't able
to handle this case with matching lifetimes against variants either.
```bash
 1  error[E0392]: parameter `'b` is never used
    --> borsh/tests/test_schema_enums.rs:218:31
 2  error[E0392]: parameter `'a` is never used
    --> borsh/tests/test_schema_enums.rs:218:27
```

File: .github/workflows/rust.yml [hunk 2]
```diff
@@ -69,6 +69,14 @@ jobs:
     - name: Run cargo fmt
       run: cargo fmt --check
 
+  cargo-doc:
+    runs-on: ubuntu-20.04
+
+    steps:
+    - uses: actions/checkout@v2
+    - name: run cargo doc
+      run: RUSTDOCFLAGS="-D warnings" cargo doc --features derive,schema
```
Comment (dj8yfo) [inline .github/workflows/rust.yml:78]: added this check to ci, as there're sometimes warnings, which signal, that the generated doc
won't look as expected, when rendered, and sometimes errors, which will prevent doc from being 
generated at release time altogether.

Comment (frol) [reply .github/workflows/rust.yml:78]: This is a valuable piece of context worth leaving a comment in the file itself, so we don't lose it

File: borsh/tests/test_generic_struct.rs [hunk 3]
```diff
@@ -133,6 +169,17 @@ where
...
 }
 
+#[allow(unused)]
+#[derive(BorshSerialize)]
+struct ParametrizedWronDerive<T, V>
+where
+    T: TraitName,
+{
+    #[borsh(bound(serialize = "<T as TraitName>::Associated: borsh::ser::BorshSerialize"))]
+    field: <T as TraitName>::Associated,
```
Comment (frol) [inline borsh/tests/test_generic_struct.rs:179]: `serde` does not do this magic, it just requires the developer to explicitly state the bounds:

```rust
trait TraitName {
    type Associated;
}

#[derive(serde::Serialize)]
struct ParametrizedWronDerive<T, V>
where
    T: TraitName,
    // <T as TraitName>::Associated: serde::Serialize // without this line it has a compilation error
{
    field: <T as TraitName>::Associated,
    another: V,
}
```

image

I think it is a sane way to implement it - let it raise compilation error and require the developer to explicitly state the bound. What do you think?

**Review by frol — CHANGES_REQUESTED** [review]

File: borsh/tests/test_generic_struct.rs (hunk 3 above)
Comment (frol) [reply borsh/tests/test_generic_struct.rs:179]: Ok, I see that serde also has [`#[serde(bound)]` syntax](https://serde.rs/field-attrs.html#bound).

P.S. @dj8yfo mentioned to me offline that it is useful when the struct is for all types of parameters and we don't want to restrict it to be borsh-serializable.

**Review by frol — APPROVED** [review]

**Pull request merged**